File: Python/Relativity_Modules_SSS_mods/Relativity_Modules_SSS_mod6.0_just__solve_ivp/scenarios/lambda_L_L_M/f_functions.py
from .variables import *


# def f(R):
#     return R - (lambda_staro * R_star) * (1 - (1 + (R / R_star) ** 2) ** (-beta))
# 
# def f1(R):
#     return 1 - ( (2 * lambda_staro * beta) / R_star) * (R * ((1 + (R / R_star) ** 2) ** (-beta - 1)))
# 
# def f2(R):
#     return - ((2 * lambda_staro * beta) / R_star) * (((1 + (R / R_star) ** 2) ** (-beta - 1)) * ((1 + 2 * (R / R_star) ** 2) * (-beta - 1) * ((1 + (R / R_star) ** 2) ** (-1))))
# 
# def f3(R):
#     return - ( (4 * lambda_staro * beta) / (R_star ** 3)) * (-beta - 1) * ((1 + (R / R_star) ** 2) ** (-beta - 2)) * (1 + (2 * R + 2 * ((R ** 3) / (R_star ** 2)) * (-beta - 2) * ((1 + (R / R_star) ** 2) ** (-1))))
# 
# def f32(R):
#     return f3(R) / f2(R)
# 
# # though we might need to check if we derived correctly...
# # so we'll use this other attempt


def f(R):
    return R - (lambda_staro * R_star) * (1 - (1 + (R / R_star) ** 2) ** (-beta))


def f1(R):
    """
    First derivative f'(R).
    """
    R_tilde = R / R_star
    return 1 - 2 * lambda_staro * beta * R_tilde * (1 + R_tilde ** 2) ** (-beta - 1)


def f2(R):
    """
    Second derivative f''(R).
    """
    R_tilde = R / R_star
    factor = -2 * lambda_staro * beta / R_star

    # Derivative of [R * (1 + (R/R_star)²)^(-beta - 1)] via product rule
    # We'll store the "bracket" term separately for clarity:
    bracket = (1 + R_tilde ** 2) + 2 * (-beta - 1) * R * R_tilde / R_star

    return factor * ((1 + R_tilde ** 2) ** (-beta - 2) * bracket)


def f3(R):
    """
    Third derivative f'''(R).
    """
    R_tilde = R / R_star

    # Let A(R) = (1 + R_tilde²)^(-beta - 2)
    #     B(R) = (1 + R_tilde²) + 2*(-beta - 1)*R*R_tilde/R_star
    A = (1 + R_tilde ** 2) ** (-beta - 2)
    B = (1 + R_tilde ** 2) + 2 * (-beta - 1) * R * R_tilde / R_star

    # A'(R) = derivative of (1 + R_tilde²)^(-beta - 2)
    A_prime = -2 * (beta + 2) * R_tilde / R_star * (1 + R_tilde ** 2) ** (-beta - 3)

    # B'(R) = derivative of (1 + R_tilde²) + 2*(-beta - 1)*R*R_tilde/R_star
    B_prime = -2 * (2 * beta + 1) * R / (R_star ** 2)

    # Combine via product rule
    g2 = A_prime * B + A * B_prime

    factor = -(2 * lambda_staro * beta / R_star)
    return factor * g2


def f32(R):
    return f3(R) / f2(R)

File: Julia/[0] relativity in Julia/scenarios/lambda_L_L_M/f_functions.jl
module FFunctions

using ..Variables  # Use two dots to move up one level

export f, f1, f2, f3, f32

function f(R)
    return R - (Variables.lambda_staro * Variables.R_star) * (1 - (1 + (R / Variables.R_star)^2)^(-Variables.beta))
end

function f1(R)
    """
    First derivative f'(R).
    """
    R_tilde = R / Variables.R_star
    return 1 - 2 * Variables.lambda_staro * Variables.beta * R_tilde * (1 + R_tilde^2)^(-Variables.beta - 1)
end

function f2(R)
    """
    Second derivative f''(R).
    """
    R_tilde = R / Variables.R_star
    factor = -2 * Variables.lambda_staro * Variables.beta / Variables.R_star
    bracket = (1 + R_tilde^2) + 2 * (-Variables.beta - 1) * R * R_tilde / Variables.R_star
    return factor * ((1 + R_tilde^2)^(-Variables.beta - 2) * bracket)
end

function f3(R)
    """
    Third derivative f'''(R).
    """
    R_tilde = R / Variables.R_star
    A = (1 + R_tilde^2)^(-Variables.beta - 2)
    B = (1 + R_tilde^2) + 2 * (-Variables.beta - 1) * R * R_tilde / Variables.R_star
    A_prime = -2 * (Variables.beta + 2) * R_tilde / Variables.R_star * (1 + R_tilde^2)^(-Variables.beta - 3)
    B_prime = -2 * (2 * Variables.beta + 1) * R / (Variables.R_star^2)
    g2 = A_prime * B + A * B_prime
    factor = -(2 * Variables.lambda_staro * Variables.beta / Variables.R_star)
    return factor * g2
end

function f32(R)
    return f3(R) / f2(R)
end

end  # module FFunctions
